gms/gossiper.cc

```diff
@@ -1763,38 +1763,84 @@ future<> gossiper::start_gossiping(int generation_nbr, std::map<application_stat
     });
 }
 
-future<> gossiper::do_shadow_round() {
-    return seastar::async([this, g = this->shared_from_this()] {
-        build_seeds_list();
-        _in_shadow_round = true;
-        auto t = clk::now();
-
-        // When peer node receives a syn message, it will send back a ack message.
-        // So, we need to register gossip message handlers before sending syn message.
+future<> gossiper::do_shadow_round(std::unordered_set<gms::inet_address> nodes) {
+    return seastar::async([this, g = this->shared_from_this(), nodes = std::move(nodes)] () mutable {
         get_gossiper().invoke_on_all([] (gossiper& g) {
             return g.init_messaging_service_handler();
         }).get();
 
-        while (this->_in_shadow_round) {
-            // send a completely empty syn
-            for (inet_address seed : _seeds) {
-                utils::chunked_vector<gossip_digest> digests;
-                gossip_digest_syn message(get_cluster_name(), get_partitioner_name(), digests);
-                auto id = get_msg_addr(seed);
-                logger.trace("Sending a GossipDigestSyn (ShadowRound) to {} ...", id);
-                // Do it in the background.
-                (void)ms().send_gossip_digest_syn(id, std::move(message)).handle_exception([id] (auto ep) {
-                    logger.trace("Fail to send GossipDigestSyn (ShadowRound) to {}: {}", id, ep);
-                });
+        nodes.erase(get_broadcast_address());
+
+        gossip_get_endpoint_states_request request{{
+            gms::application_state::STATUS,
+            gms::application_state::HOST_ID,
+            gms::application_state::TOKENS,
+            gms::application_state::SUPPORTED_FEATURES}};
+        logger.info("The do_shadow_round started, nodes={}", nodes);
+        std::unordered_set<gms::inet_address> node_talked;
+        size_t node_down = 0;
+        auto start_time = clk::now();
+        bool fall_back_to_syn_msg = false;
+        for (;;) {
+            for (const auto& node : nodes) {
+                logger.debug("Sent get_endpoint_states request to {}", node);
+                try {
+                    auto response = ms().send_gossip_get_endpoint_states(msg_addr(node), request).get0();
```

Comment:
We should contact nodes in parallel. If we have 100 nodes in a remote data center with 100 ms latency, we get 10 seconds delay.

Reply:
I wanted to get less contentions when applying the results. In most cases, the response from first node is the latest info, we will check and ignore the rest of 99 responses  by looking at the version. 

Since we can handle such parallel apply, talking to nodes in parallel has more advantages.  I will change it.